### PR TITLE
Task/adq 23 spike patient matching

### PR DESCRIPTION
**Pull request by ilaflamme-va**
- Framework for providing the _X-VA-INCLUDES-ICN_ header in outbound requests.

- Initial Implementation is on the Patient resource.

- All patient reads and searches will include the patient's icn as a response header. This header will be post-processed by Kong for patient-matching validation in a latter PR.

- Validation tooling for unit testing

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    implements ResponseBodyAdvice<Object> {
+
+  private Class<T> type;
+  private Class<B> bundleType;
+  private Function<T, Stream<String>> extractIcns;
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Object beforeBodyWrite(
+      Object o,
```
Comment (bschofield-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java]: maybe a better name than `o`. _What_ is the `o` parameter and _how_ is it used?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * supplied ICN extraction function. We then populate an internal header of X-VA-INCLUDES-ICN with
+ * the corresponding ICN(s) in the payload. This header will be used by Kong to do Authorization via
+ * Patient Matching.
+ */
+@AllArgsConstructor
+public abstract class AbstractIncludesIcnMajig<
+        T extends Resource, E extends AbstractEntry<T>, B extends AbstractBundle<E>>
+    implements ResponseBodyAdvice<Object> {
+
+  private Class<T> type;
```
Comment (bschofield-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java]: These can also be `final`

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajigTest.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    assertThat(new FakeMajg().supports(supportedResource, null)).isTrue();
+    MethodParameter supportedResourceBundle = mock(MethodParameter.class);
+    doReturn(FakeBundle.class).when(supportedResourceBundle).getParameterType();
+    assertThat(new FakeMajg().supports(supportedResourceBundle, null)).isTrue();
+    MethodParameter unsupportedResource = mock(MethodParameter.class);
+    doReturn(String.class).when(unsupportedResource).getParameterType();
+    assertThat(new FakeMajg().supports(unsupportedResource, null)).isFalse();
+  }
+
+  @Test(expected = InvalidParameterException.class)
```
Comment (bschofield-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajigTest.java]: It's not exactly obvious what the _intention_ is for this test. This test could be named a little better to _explain_ what it is trying to validate, e.g. `beforeBodyWriteThrowsExceptionForUnsupportedType`.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java (hunk 1 above)
Comment (ilaflamme-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java]: `payload`, now.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java (hunk 2 above)
Comment (ilaflamme-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajig.java]: done

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajigTest.java (hunk 3 above)
Comment (ilaflamme-va) [reply data-query/src/test/java/gov/va/api/health/dataquery/service/controller/AbstractIncludesIcnMajigTest.java]: done

**Pull request merged**